### ISSUE TITLE
Higher-level api for defining java functions in a more "rusty" manner

### ISSUE THREAD

**Issue by jrobsonchase** [jni-rs/jni-rs#80]
See #76 for some background.

It would be nice to be able to define the FFI functions more like regular rust functions with rust error handling rather than as `extern "C"` functions. The current way of doing things forces you to manually catch `Result::Err` results and turn them into java exceptions (or however you want to deal with them), and to manually wrap everything in `std::panic::catch_unwind` (which you are doing, right?) and deal with the `Result` from that as well.

We could instead have a macro system that allows you to write your ffi methods such that they return the expected `Result<T, E>`, and then convert that into an exception or java return type as needed. It would do the same with automatically doing the `catch_unwind` wrapping and result management in a similar manner.

The macro system could potentially also do automatic conversions of java types (such as `String`s) into their rust representations, though it should probably be possible to bypass this if desired for performance reasons.

**Comment by mehcode** [comment]
Definitely want to see this. Something like:

```rust
#[jni("com.example.Foo")]
pub fn init(filename: &str) -> Result<(), Error> { ... }

#[jni("com.example.Bar")]
pub fn close_it(env: JNIEnv, filename: JString) -> Result<(), Error> { ... }
```

Becomes something like 

```rust
#[no_mangle]
pub extern "system" fn Java_com_example_Bar_closeIt(env: JNIEnv, _: JClass, filename: JString) {
  // [...]
} 
```

1. Name maker. Pass in java path and the name becomes java-y. 
2. Auto-conversion between Rust and Java types. Java types can still be expressed if you want to do it yourself.
3. `catch_unwind`
4. Handle `Result`s and throw on error

**Comment by Progdrasil** [comment]
After having used a functionality pretty muchg exactly as @mehcode describes in Pyo3, this would be a great productivity boost.

**Comment by Diegovsky** [comment]
Agreed. Having to mix java's exceptions and rust's `Result` enum, I'm confident the syntax proposed would look and feel so much better.

**Comment by antonok-edm** [comment]
After playing around with some JNI bindings of my own and encountering similar pain points, I've just published [`jni_fn`](https://crates.io/crates/jni_fn) which accomplishes at least the first item from [@mehcode's comment](https://github.com/jni-rs/jni-rs/issues/80#issuecomment-378136868):

> 1. Name maker. Pass in java path and the name becomes java-y.

That seemed like the most well-defined problem to knock off the list, and was also my biggest confusion when starting out (where did this long auto-generated name come from and why do I need to care about it?).

I'm happy to explore incorporating the `jni_fn` macro into `jni` if there's interest!

I'm also open to adding in the other 3 items from that list, but I'd appreciate more design discussion; ideally examples of code before and after expansion. I've only just started investigating JNI and I'm sure I don't have a good mental model of the "edge cases" of JNI usage yet.

**Comment by haraldreingruber-dedalus** [comment]
> After playing around with some JNI bindings of my own and encountering similar pain points, I've just published [`jni_fn`](https://crates.io/crates/jni_fn) which accomplishes at least the first item from [@mehcode's comment](https://github.com/jni-rs/jni-rs/issues/80#issuecomment-378136868):
> 
> > 1. Name maker. Pass in java path and the name becomes java-y.
> 
...
> I'm happy to explore incorporating the `jni_fn` macro into `jni` if there's interest!
> 

@antonok-edm  That looks super useful! Will try it in our project soon. Thanks for sharing 🙏

**Comment by uvlad7** [comment]
This topic is pretty old, but maybe someone will find this useful: I created [macros](https://github.com/uvlad7/xml2json-rb/blob/0e794a5e5b12632a6a3c69b96147cf5955ff80e0/ext/xml2json/jagnus/src/jni_function.rs) to handle exactly what is described here. It's inspired by [magnus](https://github.com/matsadler/magnus/blob/c2e125e140e1903777ba369f5eabe1b71561d573/src/method.rs) and it's WIP, no tests, no `jni[_static]_method` implementation, poor `Err` handling.

Also you may find this [project](https://github.com/giovanniberti/robusta) useful, though I don't like the approach a bit and think PyO3's and Magnus' way is better.

**Comment by wuwbobo2021** [comment]
This issue might be closed now, see <https://docs.rs/jni/0.22.1/jni/macro.native_method.html>.

**Comment by rib** [comment]
> This issue might be closed now, see https://docs.rs/jni/0.22.1/jni/macro.native_method.html.

yes, I think [`#[jni_mangle]`](https://github.com/jni-rs/jni-rs/blob/master/crates/jni/docs/macros/jni_mangle.md) (similar to `jni_fn` mentioned above),  [`native_method!`](https://docs.rs/jni/latest/jni/macro.native_method.html) and [`bind_java_type!`](https://docs.rs/jni/latest/jni/macro.bind_java_type.html) hopefully all address this issue.

In particular `native_method!` and `bind_java_type!` build on the [`jni_sig!`](https://docs.rs/jni/latest/jni/macro.jni_sig.html) syntax that allows for more natural signatures based on Rust and Java types instead of the more-esoteric JNI signature strings (though these are also still supported).

Incidentally, I just realised that the docs for `jni_mangle` aren't being properly pulled in to the `jni` crate reference docs, like they are for the other macros.

Specifically in regards to importing the `jni_fn` crate - that's essentially what the `jni_mangle` attribute macros is. It was originally based on an import of `jni_fn` and then extended to:
- more-fully support the mangling rules, e.g. for unicode and overloaded JNI methods
- support use of `snake_case` names being automatically mapped to `lowerCamelCase`
- use `export_name` internally instead of actually renaming the Rust function - so your rust function remains accessible via it's original, un-mangled name

Please re-open if we don't think this has been addressed
